Make Teams module pure with addition of a seed value for 'random'

sarahabimay opened the pull request:
**Remove a side_cause from a method, in the form of global random seed**
RandomTeamAllocator::assign_teams was calling :rand.export_seed() in order to send back that seed to the TeamBuilderApp ready for the next time 'build teams' command was called.  This is a 'side cause', aka unknown input and should be avoided in order to make the function pure.
I moved it to the point where it actually changes.
This has the effect of complicating the return value a little as the return value is mixed and looks a little like this:
[%{team => 1, member => "a"}, %{:new_seed => {:exsplus, [111111 | 22222] }]
I'm not really happy with this so I will come back to it as it has taken 2 points to get this far.

@maikon
@jsuchy
@felipesere

**Previous commit's comments:**
I realised that I was using the Erlang 'rand' module and that I had made my modules impure by proxy.
I wanted to verify that I was right so I paired with Mateu and he confirmed my suspicions.  He helped me by explaining what the 'seed' part of :rand in Elixir (and similar in Ruby) and how if I knew what the seed was when my module made a random selection then, then it would be predictable what order each subsequent random selection would be.
That was great, but then I realised that if the same seed value was used each time a random selection was made then that meant the same order of random choices would be re-played.  This obviously doesn't look random if the member set remains the same.  I knew that I would need to create a new seed  each time a 'build teams' request took place, but I couldn't figure out how to do this and keep it pure.  I needed to work it through with Mateu as it was fairly unintuitive for me.  
I haven't checked the final code with him, but will ask him to look over it.

sarahabimay committed:
Introduce random team allocation

@maikon
@jsuchy
@felipeser

sarahabimay committed:
Refactor Teams module to aid readability

sarahabimay committed:
Make Teams module pure with addition of a seed value for 'random'

sarahabimay committed:
Each 'build teams' command now will uses a different seed.

@maikon
@jsuchy
@felipesere

I realized that I was using the same seed each time a 'build teams' request
was received and that meant the same order of random choices was being re-played.  This obviously doesn't look random if the member set remains the same.  It took me a while to work out how to do this.  I needed to work it through with Mateu as it's a fairly complicated topic.

sarahabimay committed:
:rand method name was incorrect in 'main'

mateuadsuara commented on lib/random_team_allocator.ex:
```diff
@@ -0,0 +1,46 @@
+
+  defp take_random_pure(members, team_count, seed) do
+    :rand.seed(seed)
+    Enum.take_random(members, team_count)
```
Now `take_random_pure` is pure in the sense of "will return the same result if the same arguments are passed". That is good :)

But it still is doing a side effect that someone else is observing (depending upon). Specifically, `take_random_pure` is still changing the global seed (because it uses `Enum.take_random` which does that) that `assign_teams` is depending upon with `:rand.export_seed()` in order to be passed at some point again into `take_random_pure`.

In a sense, it is not having the details at the same level of abstraction.
Specifically: `:rand.seed(seed)` and `:rand.export_seed()` are at the same level of abstraction, but they are placed at different levels of abstraction in `TeamBuilder`: `take_random_pure` which is using `:rand.seed(seed)` is very abstract (general use) and `assign_teams` which is using `:rand.export_seed()` is very concrete (specific to this app).

I'll elaborate more if it is difficult to see how to solve that. :)

Reference: https://en.wikipedia.org/wiki/Pure_function

sarahabimay committed:
Remove a side_cause from a method, in the form of global random seed
RandomTeamAllocator::assign_teams was calling :rand.export_seed() in order to send back that seed to the TeamBuilderApp ready for the next time 'build teams' command was called.  This is a 'side cause', aka unknown input and should be avoided in order to make the function pure.
I moved it to the point where it actually changes.
This has the effect of complicating the return value a little as the return value is mixed and looks a little like this:
[%{team => 1, member => "a"}, %{:new_seed => {:exsplus, [111111 | 22222] }]
I'm not really happy with this so I will come back to it as it has taken 2 points to get this far.
@maikon
@jsuchy
@felipesere

sarahabimay committed:
Change parameter name
